base/compiler/tfuncs.jl

```diff
@@ -1464,7 +1452,24 @@ function builtin_tfunction(interp::AbstractInterpreter, @nospecialize(f), argtyp
         tf = T_FFUNC_VAL[fidx]
     end
     tf = tf::Tuple{Int, Int, Any}
-    if !(tf[1] <= length(argtypes) <= tf[2])
+    if !isempty(argtypes) && isvarargtype(argtypes[end])
+        if length(argtypes) - 1 > tf[2]
+            # definitely too many arguments
+            return Bottom
+        end
+        if length(argtypes) - 1 == tf[2]
+            argtypes = argtypes[1:end-1]
+        else
+            vatype = argtypes[end]
+            argtypes = argtypes[1:end-1]
+            while length(argtypes) < tf[1]
+                push!(argtypes, unwrapva(vatype))
+            end
+            if length(argtypes) < tf[2]
+                push!(argtypes, vatype)
```

Comment:
possibly might also need to adjust a length constraint on vatype (though I don't know if we'd ever form this)? for example, if it was `Vararg{NTuple{N}, N} where N`. This is probably already broken elsewhere, so not to worry now

Reply:
Good catch! Even if broken elsewhere, no excuse for not handling it correctly here, even if it doesn't make much difference now. As long as it's not overly complicated, that is, but I think this should do it:
```suggestion
                push!(argtypes, unconstrain_vararg_length(vatype))
```
There may be corner case where we loose a little bit of precision due to this, but that's better than giving a wrong result. Or am I missing something and this simple fix won't do it?